# Выпускной проект - SQL

## Описание проекта

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

### Описание данных

![jupyter](https://concrete-web-bad.notion.site/image/https%3A%2F%2Fs3-us-west-2.amazonaws.com%2Fsecure.notion-static.com%2F069818d1-0e5c-4d87-a461-0de584ab9c33%2FUntitled_(33).png?table=block&id=bd53c8db-b4fd-49eb-8cc9-572ebb3c9163&spaceId=9e4bd47b-c6e6-4ca3-bcee-279794b47315&width=2000&userId=&cache=v2)

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

### Задания

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

## Подключение к БД

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

In [3]:
query_books = ''' SELECT *
                  FROM books;
              '''

In [4]:
books = pd.io.sql.read_sql(query_books, con = engine) 
books.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [5]:
query_authors = ''' SELECT *
                    FROM authors;
                '''

In [6]:
authors = pd.io.sql.read_sql(query_authors, con = engine) 
authors.head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [7]:
query_publishers = ''' SELECT *
                       FROM publishers;
                   '''

In [8]:
publishers = pd.io.sql.read_sql(query_publishers, con = engine) 
publishers.head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [9]:
query_ratings = ''' SELECT *
                    FROM ratings;
                '''

In [10]:
ratings = pd.io.sql.read_sql(query_ratings, con = engine) 
ratings.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [11]:
query_reviews = ''' SELECT *
                    FROM reviews;
                '''

In [12]:
reviews = pd.io.sql.read_sql(query_reviews, con = engine) 
reviews.head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [13]:
# ячейка ревьюера: загружаем енсколько строк:
display(pd.io.sql.read_sql(
    ''' SELECT *
    FROM books
    LIMIT 1
    '''
, con = engine)) 

# смотрим на объем данных:
display(pd.io.sql.read_sql(
    ''' SELECT COUNT(*)
    FROM books
    LIMIT 1
    '''
, con = engine))                

book_id  author_id         title  num_pages publication_date  publisher_id
0        1        546  'Salem's Lot        594       2005-11-01            93

count
0   1000

In [14]:
# можно подсчитать количество строк сразу с использованием небольшой оконной функции:

pd.io.sql.read_sql('''
SELECT *,
       COUNT(*) OVER() AS table_size
FROM books 
LIMIT 1        
'''
, con = engine)

book_id  author_id         title  num_pages publication_date  publisher_id  \
0        1        546  'Salem's Lot        594       2005-11-01            93   

   table_size  
0        1000

### Задача №1

- **Посчитайте, сколько книг вышло после 1 января 2000 года**

In [15]:
sql_1 = '''SELECT COUNT(book_id)
           FROM books
           WHERE publication_date > '2000-01-01';
        '''

pd.read_sql(sql_1, con = engine)

count
0    819

### Вывод: 

- **После 1 января 2000 года вышло 819 книг.**

### Задача №2
- **Для каждой книги посчитайте количество обзоров и среднюю оценку**

In [16]:
sql_2 = '''SELECT books.book_id,
                  books.title, 
                  COUNT(DISTINCT reviews.review_id) AS count_review,  
                  AVG(ratings.rating) AS avg_ratings         
           FROM books
           LEFT JOIN ratings ON ratings.book_id = books.book_id 
           LEFT JOIN reviews ON reviews.book_id = books.book_id 
           GROUP BY books.book_id
           ORDER BY count_review DESC,avg_ratings DESC;
        '''   

pd.read_sql(sql_2, con = engine)

book_id                                              title  count_review  \
0        948                            Twilight (Twilight  #1)             7   
1        302  Harry Potter and the Prisoner of Azkaban (Harr...             6   
2        299  Harry Potter and the Chamber of Secrets (Harry...             6   
3        656                                     The Book Thief             6   
4        734                                   The Glass Castle             6   
..       ...                                                ...           ...   
995      191  Disney's Beauty and the Beast (A Little Golden...             0   
996      387                               Leonardo's Notebooks             0   
997      221                          Essential Tales and Poems             0   
998       83    Anne Rice's The Vampire Lestat: A Graphic Novel             0   
999      808                            The Natural Way to Draw             0   

     avg_ratings  
0       3.662500  
1       4.414634  
2       4.287500  
3       4.264151  
4       4.206897  
..           ...  
995     4.000000  
996     4.000000  
997     4.000000  
998     3.666667  
999     3.000000  

[1000 rows x 4 columns]

### Вывод: 

- **Максимальное количество обзоров у Twilight(Twilight#1) -7, средняя оценка - 3.6625.**

### Задача №3

- **Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры**

In [17]:
sql_3 = '''SELECT publishers.publisher,
                  COUNT(books.book_id) AS count_books
           FROM publishers
           LEFT JOIN books ON publishers.publisher_id = books.publisher_id
           WHERE num_pages > 50
           GROUP BY publishers.publisher
           ORDER BY count_books DESC
           LIMIT 1;
        '''   

pd.read_sql(sql_3, con = engine)

publisher  count_books
0  Penguin Books           42

### Вывод: 

- **Издательство Penguin Books выпустило 42 книги толще 50 страниц.**

### Задача №4
- **- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками**

In [18]:
sql_4 = '''SELECT authors.author,
                  AVG(ratings.rating) AS avg_rat,
                  COUNT(ratings.rating) AS count_rat
           FROM authors
           LEFT JOIN books ON authors.author_id = books.author_id
           LEFT JOIN ratings ON books.book_id = ratings.book_id
           WHERE books.book_id IN
                 (SELECT books.book_id
                 FROM books    
                 LEFT JOIN ratings ON ratings.book_id = books.book_id     
                 GROUP BY books.book_id
                 HAVING COUNT(ratings.rating_id) >= 50)
           GROUP BY authors.author  
           ORDER BY avg_rat DESC
           LIMIT 1;
        '''   

pd.read_sql(sql_4, con = engine)

author   avg_rat  count_rat
0  J.K. Rowling/Mary GrandPré  4.287097        310

### Вывод: 
-  **J.K. Rowling - автор с самой высокой средней оценкой книг - 4.88.**

In [19]:
query = ''' 
    WITH book_rating AS
      (SELECT book_id,
              AVG(rating) AS avg_rating
       FROM ratings
       WHERE book_id IN
           (SELECT book_id
            FROM ratings
            GROUP BY book_id
            HAVING COUNT(rating_id) >=50)
       GROUP BY book_id)
    SELECT a.author AS author,
           AVG(book_rating.avg_rating) AS best_raiting
    FROM authors AS a
    JOIN books AS b ON b.author_id = a.author_id
    JOIN book_rating ON book_rating.book_id = b.book_id
    GROUP BY a.author_id,
             a.author
    ORDER BY best_raiting DESC
        '''
pd.io.sql.read_sql(query, con = engine)


author  best_raiting
0                          J.K. Rowling/Mary GrandPré      4.283844
1                   Markus Zusak/Cao Xuân Việt Khương      4.264151
2                                      J.R.R. Tolkien      4.258446
3                                   Louisa May Alcott      4.192308
4                                        Rick Riordan      4.080645
5                                     William Golding      3.901408
6                                       J.D. Salinger      3.825581
7            Paulo Coelho/Alan R. Clarke/Özdemir İnce      3.789474
8   William Shakespeare/Paul Werstine/Barbara A. M...      3.787879
9                                           Dan Brown      3.754540
10                                         Lois Lowry      3.750000
11            George Orwell/Boris Grabnar/Peter Škerl      3.729730
12                                    Stephenie Meyer      3.662500
13                                     John Steinbeck      3.622951

### Задача №5

- **Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок**

In [20]:
sql_5 = '''WITH
               tab_ratings AS 
               (SELECT ratings.username,
                       COUNT(rating_id) AS count_ratings
                FROM ratings
                GROUP BY ratings.username
                HAVING COUNT(rating_id) > 50),
               tab_reviews AS 
               (SELECT reviews.username,
                       COUNT(review_id) AS count_reviews
                FROM reviews
                GROUP BY reviews.username)
           SELECT AVG(count_reviews) as avg_count_reviews
           FROM tab_ratings
           LEFT JOIN tab_reviews ON tab_ratings.username = tab_reviews.username
        '''

pd.read_sql(sql_5, con = engine)

avg_count_reviews
0          24.333333

### Вывод:
- **среднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24.**

## Общие выводы:

**В ходе проекта мы проанализировали базу данных крупного сервиса для чтения книг**


- После 1 января 2000 года было выпущено 819 книг;
- Максимальное количество обзоров у Twilight(Twilight#1) -7, со средней оценкой - 3.66;
- Наибольшее число книг толще 50 страниц выпустило издательство **Penguin Books** - 42 книги;
- Самые высокие оценки у автора Д. Роулинг - 4.28 балла;
- Среднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24.